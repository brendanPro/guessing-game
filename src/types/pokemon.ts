export interface Pokemon {
  id: number;
  name: string;
  frenchName: string;
  sprites: {
    front_default: string;
    back_default: string;
    front_shiny: string;
    back_shiny: string;
  };
  types: Array<{
    type: {
      name: string;
    };
  }>;
  height: number;
  weight: number;
  stats: Array<{
    base_stat: number;
    stat: {
      name: string;
    };
  }>;
}

export interface PokemonSpecies {
  count: number;
  results: Array<{
    name: string;
    url: string;
  }>;
}

export enum GameStatus {
  PLAYING = 'playing',
  WON = 'won',
  LOST = 'lost',
}

export enum GameMode {
  BLUR = 'blur',
  ZOOM = 'zoom',
}

export interface GameState {
  targetPokemon: Pokemon | null;
  attempts: string[];
  currentAttempt: string;
  status: GameStatus;
  blurLevel: number; // 0-5, where 0 is fully blurred/zoomed, 5 is clear/normal
  gameMode: GameMode;
}

export enum LetterState {
  CORRECT = 'correct',
  PRESENT = 'present',
  ABSENT = 'absent',
}
