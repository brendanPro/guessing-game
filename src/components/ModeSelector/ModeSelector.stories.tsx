import "@/index.css"
import type { Meta, StoryObj } from '@storybook/react';
import { ModeSelector } from './ModeSelector';
import { GameMode } from '@/types/pokemon';
import { fn } from '@storybook/test';
import { useState } from 'react';

const meta = {
  title: 'Game/ModeSelector',
  component: ModeSelector,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    currentMode: {
      control: 'select',
      options: [GameMode.BLUR, GameMode.ZOOM],
      description: 'The currently selected game mode',
    },
    onModeChange: {
      action: 'mode changed',
      description: 'Callback when mode is changed',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the selector is disabled',
    },
  },
  args: {
    onModeChange: fn(),
  },
} satisfies Meta<typeof ModeSelector>;

export default meta;
type Story = StoryObj<typeof meta>;

export const BlurMode: Story = {
  args: {
    currentMode: GameMode.BLUR,
    disabled: false,
  },
};

export const ZoomMode: Story = {
  args: {
    currentMode: GameMode.ZOOM,
    disabled: false,
  },
};

export const Disabled: Story = {
  args: {
    currentMode: GameMode.BLUR,
    disabled: true,
  },
};

export const Interactive: Story = {
  args: {
    currentMode: GameMode.BLUR,
    disabled: false,
  },
  render: (args) => {
    const [mode, setMode] = useState(args.currentMode);
    
    return (
      <div className="space-y-4">
        <ModeSelector
          {...args}
          currentMode={mode}
          onModeChange={(newMode) => {
            setMode(newMode);
            args.onModeChange?.(newMode);
          }}
        />
        <div className="text-center p-4 bg-muted rounded-lg">
          <p className="text-sm">
            <strong>Current Mode:</strong> {mode === GameMode.BLUR ? 'Mode Flou' : 'Mode Zoom'}
          </p>
          <p className="text-xs text-muted-foreground mt-2">
            Use the dropdown to switch between modes
          </p>
        </div>
      </div>
    );
  },
};

export const ModeSwitching: Story = {
  args: {
    currentMode: GameMode.BLUR,
    disabled: false,
  },
  render: (args) => {
    const [mode, setMode] = useState(args.currentMode);
    
    return (
      <div className="space-y-4 w-full max-w-md">
        <div className="text-center mb-4">
          <h3 className="text-lg font-semibold mb-2">Mode Selector Demo</h3>
          <p className="text-sm text-muted-foreground">
            Use the dropdown to select between Blur and Zoom modes
          </p>
        </div>
        <ModeSelector
          {...args}
          currentMode={mode}
          onModeChange={(newMode) => {
            setMode(newMode);
            args.onModeChange?.(newMode);
          }}
        />
        <div className="grid grid-cols-2 gap-4 text-center">
          <div className={`p-4 rounded-lg border-2 ${mode === GameMode.BLUR ? 'border-primary bg-primary/10' : 'border-border'}`}>
            <p className="font-semibold">Mode Flou</p>
            <p className="text-xs text-muted-foreground mt-1">Image progressively unblurs</p>
          </div>
          <div className={`p-4 rounded-lg border-2 ${mode === GameMode.ZOOM ? 'border-primary bg-primary/10' : 'border-border'}`}>
            <p className="font-semibold">Mode Zoom</p>
            <p className="text-xs text-muted-foreground mt-1">Image progressively zooms out</p>
          </div>
        </div>
      </div>
    );
  },
};

