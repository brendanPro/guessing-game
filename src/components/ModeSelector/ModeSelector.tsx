import React from 'react';
import { Card, CardContent } from '@/components/ui/card';
import { Button } from '@/components/ui/button';
import { cn } from '@/lib/utils';
import { GameMode } from '@/types/pokemon';
import { ChevronLeft, ChevronRight } from 'lucide-react';

interface ModeSelectorProps {
  currentMode: GameMode;
  onModeChange: (mode: GameMode) => void;
  disabled?: boolean;
  className?: string;
}

const MODE_NAMES: Record<GameMode, string> = {
  [GameMode.BLUR]: 'Mode Flou',
  [GameMode.ZOOM]: 'Mode Zoom',
};

const MODES = [GameMode.BLUR, GameMode.ZOOM];

export function ModeSelector({ 
  currentMode, 
  onModeChange, 
  disabled = false,
  className 
}: ModeSelectorProps) {
  const currentIndex = MODES.indexOf(currentMode);

  const handlePrevious = () => {
    const newIndex = currentIndex === 0 ? MODES.length - 1 : currentIndex - 1;
    onModeChange(MODES[newIndex]);
  };

  const handleNext = () => {
    const newIndex = currentIndex === MODES.length - 1 ? 0 : currentIndex + 1;
    onModeChange(MODES[newIndex]);
  };

  return (
    <Card className={cn("w-full max-w-md mx-auto", className)}>
      <CardContent className="p-3 sm:p-4">
        <div className="flex items-center justify-between gap-2">
          <Button
            variant="outline"
            size="icon"
            onClick={handlePrevious}
            disabled={disabled}
            className="shrink-0"
            aria-label="Mode précédent"
          >
            <ChevronLeft className="h-4 w-4" />
          </Button>
          
          <div className="flex-1 text-center">
            <span className="text-sm sm:text-base font-semibold">
              {MODE_NAMES[currentMode]}
            </span>
          </div>
          
          <Button
            variant="outline"
            size="icon"
            onClick={handleNext}
            disabled={disabled}
            className="shrink-0"
            aria-label="Mode suivant"
          >
            <ChevronRight className="h-4 w-4" />
          </Button>
        </div>
      </CardContent>
    </Card>
  );
}

