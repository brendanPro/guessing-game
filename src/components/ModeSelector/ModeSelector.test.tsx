import { describe, test, expect, mock } from "bun:test";
import { renderToStaticMarkup } from "react-dom/server";
import { ModeSelector } from "./ModeSelector";
import { GameMode } from '@/types/pokemon';


describe("ModeSelector Component", () => {
  const mockOnChange = mock(() => {});

  test("snapshot with blur mode selected", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toMatchSnapshot();
  });

  test("snapshot with zoom mode selected", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.ZOOM} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toMatchSnapshot();
  });

  test("snapshot when disabled", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={true}
      />
    );
    expect(html).toMatchSnapshot();
  });

  test("displays blur mode name", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("Mode Flou");
  });

  test("displays zoom mode name", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.ZOOM} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("Mode Zoom");
  });

  test("contains left arrow button", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain('aria-label="Mode précédent"');
  });

  test("contains right arrow button", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain('aria-label="Mode suivant"');
  });

  test("has correct layout structure", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("flex");
    expect(html).toContain("items-center");
    expect(html).toContain("justify-between");
  });

  test("buttons are disabled when disabled prop is true", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={true}
      />
    );
    expect(html).toContain('disabled=""');
  });

  test("has responsive padding", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("p-3");
    expect(html).toContain("sm:p-4");
  });

  test("has centered text", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("text-center");
  });

  test("buttons have icon size", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("size-9");
  });

  test("applies custom className when provided", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
        className="custom-class"
      />
    );
    expect(html).toContain("custom-class");
  });

  test("has responsive text size", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("text-sm");
    expect(html).toContain("sm:text-base");
  });

  test("buttons have outline styling", () => {
    const html = renderToStaticMarkup(
      <ModeSelector 
        currentMode={GameMode.BLUR} 
        onModeChange={mockOnChange}
        disabled={false}
      />
    );
    expect(html).toContain("border-input");
    expect(html).toContain("bg-background");
  });
});

