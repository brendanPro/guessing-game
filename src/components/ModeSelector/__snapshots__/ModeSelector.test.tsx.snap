// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ModeSelector Component snapshot with blur mode selected 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><div class="flex items-center justify-between gap-2"><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 border border-input bg-background shadow-xs hover:bg-accent hover:text-accent-foreground size-9 shrink-0" aria-label="Mode précédent"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-left h-4 w-4"><path d="m15 18-6-6 6-6"></path></svg></button><div class="flex-1 text-center"><span class="text-sm sm:text-base font-semibold">Mode Flou</span></div><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 border border-input bg-background shadow-xs hover:bg-accent hover:text-accent-foreground size-9 shrink-0" aria-label="Mode suivant"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-right h-4 w-4"><path d="m9 18 6-6-6-6"></path></svg></button></div></div></div>"`;

exports[`ModeSelector Component snapshot with zoom mode selected 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><div class="flex items-center justify-between gap-2"><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 border border-input bg-background shadow-xs hover:bg-accent hover:text-accent-foreground size-9 shrink-0" aria-label="Mode précédent"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-left h-4 w-4"><path d="m15 18-6-6 6-6"></path></svg></button><div class="flex-1 text-center"><span class="text-sm sm:text-base font-semibold">Mode Zoom</span></div><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 border border-input bg-background shadow-xs hover:bg-accent hover:text-accent-foreground size-9 shrink-0" aria-label="Mode suivant"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-right h-4 w-4"><path d="m9 18 6-6-6-6"></path></svg></button></div></div></div>"`;

exports[`ModeSelector Component snapshot when disabled 1`] = `"<div data-slot="card" class="bg-card text-card-foreground rounded-xl border shadow-sm w-full max-w-md mx-auto"><div data-slot="card-content" class="p-3 sm:p-4"><div class="flex items-center justify-between gap-2"><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 border border-input bg-background shadow-xs hover:bg-accent hover:text-accent-foreground size-9 shrink-0" disabled="" aria-label="Mode précédent"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-left h-4 w-4"><path d="m15 18-6-6 6-6"></path></svg></button><div class="flex-1 text-center"><span class="text-sm sm:text-base font-semibold">Mode Flou</span></div><button data-slot="button" class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg:not([class*=&#x27;size-&#x27;])]:size-4 [&amp;_svg]:shrink-0 ring-ring/10 dark:ring-ring/20 dark:outline-ring/40 outline-ring/50 focus-visible:ring-4 focus-visible:outline-1 aria-invalid:focus-visible:ring-0 border border-input bg-background shadow-xs hover:bg-accent hover:text-accent-foreground size-9 shrink-0" disabled="" aria-label="Mode suivant"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-chevron-right h-4 w-4"><path d="m9 18 6-6-6-6"></path></svg></button></div></div></div>"`;
